### PR TITLE
Revert "[libc++] LWG3870: Remove `voidify` (#110355)"

### PR DESCRIPTION
**Pull request by Michael137**
This reverts commit 78f9a8b82d772ff04a12ef95f2c9d31ee8f3e409.

This caused the LLDB test `TestDataFormatterGenericOptional.py` to fail with:
```
FAIL: test_with_run_command_libcpp_dwarf (TestDataFormatterGenericOptional.GenericOptionalDataFormatterTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 1769, in test_method
    return attrvalue(self)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 148, in wrapper
    return func(*args, **kwargs)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 148, in wrapper
    return func(*args, **kwargs)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/data-formatter/data-formatter-stl/generic/optional/TestDataFormatterGenericOptional.py", line 103, in test_with_run_command_libcpp
    self.do_test_with_run_command(USE_LIBCPP)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/data-formatter/data-formatter-stl/generic/optional/TestDataFormatterGenericOptional.py", line 59, in do_test_with_run_command
    self.expect(
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 2475, in expect
    self.fail(log_msg)
AssertionError: Ran command:
"frame var numbers"

Got output:
(optional_int_vect) numbers =  Has Value=true  {
  Value = {
    __begin_ = 0x0000600002be8000
    __end_ = 0x0000600002be8010
    __cap_ = 0x0000600002be8010
    __padding1_781_ = (__padding_ = "\U00000001\0\xf7\0ה\xd2\xf0*\0\0\0*\0\0\0\U00000001\xb4Zo*\0\0\0*\0\0\0\U00000001\0\0\0\0\U0000001d\xc7\0\0\0\0\0\xbc[\x85\U00000001\0\0\0\0\U00000001\0\0\0\U00000002\0\0\0\U00000003\0\0\0\U00000004")
    __alloc_ = {}
    __padding2_781_ = (__padding_ = "\U00000001\0\xf7\0ה\xd2\xf0*\0\0\0*\0\0\0\U00000001\xb4Zo*\0\0\0*\0\0\0\U00000001\0\0\0\0\U0000001d\xc7\0\0\0\0\0\xbc[\x85\U00000001\0\0\0\0\U00000001\0\0\0\U00000002\0\0\0\U00000003\0\0\0\U00000004")
  }
}

Expecting sub string: "(optional_int_vect) numbers =  Has Value=true  {" (was found)
Expecting sub string: "Value = size=4 {" (was not found)
Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
----------------------------------------------------------------------
Ran 6 tests in 3.259s

FAILED (failures=2, skipped=4)
```

Will need a bit more time to look into a solution

### CODE REVIEW

**Comment by Michael137** [comment]
@ldionne let me know if the modulemap doesn't look right

File: libcxx/include/module.modulemap [hunk 1]
```diff
@@ -2237,3 +2237,1470 @@ module std_private_mbstate_t [system] {
...
+  export std_private_type_traits_type_identity
+}
+module std_private_memory_unique_temporary_buffer         [system] {
+  header "__memory/unique_temporary_buffer.h"
+  export std_private_memory_unique_ptr
+  export std_private_type_traits_is_constant_evaluated
+}
+module std_private_memory_uses_allocator                  [system] { header "__memory/uses_allocator.h" }
+module std_private_memory_uses_allocator_construction     [system] { header "__memory/uses_allocator_construction.h" }
+module std_private_memory_voidify                         [system] { header "__memory/voidify.h" }
```
Comment (frederick-vs-ja) [inline libcxx/include/module.modulemap:3134]: I think only `module voidify                            { header "__memory/voidify.h" }` should be added back after this line:
https://github.com/llvm/llvm-project/blob/eea5e7e095f945329e9e71c9606075264df8d74f/libcxx/include/module.modulemap#L1530

Comment (Michael137) [reply libcxx/include/module.modulemap]: ah thanks for the correction, fixed

**Review by ldionne — APPROVED** [review]
Merge conflict resolution LGTM. @Michael137 Please let us know once you've had time to look into this.

**Pull request merged**